File: Algorithm and Data Structure/python/breadth_first_value.py
class Node:
    def __init__(self, val) -> None:
        self.val = val
        self.right = None
        self.left = None

a, b, c, d, e, f = (Node('a'), Node('b'), Node('c'), 
                    Node('d'), Node('e'), Node('f'))

a.left, a.right = (b, c)
b.left, b.right = (d, e)
c.right = f

def breadth_first_value(root):
    if root == None:
        return []
    result = []
    queue = [root]

    while len(queue) > 0:
        current = queue[0]
        queue.remove(current)
        result.append(current.val)
        
        if current.left: queue.append(current.left)
        if current.right: queue.append(current.right)
    return result

print(breadth_first_value(a))
"""
        a
        /\
       b  c    Left to Right: [a b c d e f]
      /\   \   Right to Left: [a c b f e d]
     d  e   f
"""

File: js/breadth-first-value.js
import {Node} from './Node.js'

const breadthFirstValues = (root) => {
    if (root == null) { 
        return [] 
    }
    
    const result = []
    const queue = [root]

    while (queue.length > 0) {
        const current = queue.shift()
        result.push(current.val)

        if (current.left) {queue.push(current.left)}
        if (current.right) {queue.push(current.right)}
    }
    return result
}

const a = new Node('a');
const b = new Node('b');
const c = new Node('c');
const d = new Node('d');
const e = new Node('e');
const f = new Node('f');

[a.left, a.right] = [b, c];
[b.left, b.right] = [d, e];
c.right = f;


console.log(breadthFirstValues(a))
